src/transformStateWithClones.js

```diff
@@ -6,8 +6,66 @@
  *
  * @return {Object[]}
  */
-function transformStateWithClones(state, actions) {
-  // write code here
+
+const state = {
+  foo: 'bar',
+  name: 'Jim',
+  another: 'one',
+};
+
+function transformStateWithClones(currentState, actions) {
+  const history = [];
+  let clonedState = Object.assign({}, currentState);
+
+  actions.forEach((obj) => {
+    if (obj.type === 'clear') {
+      clonedState = {};
+      history.push({});
+    }
+
+    if (obj.type === 'addProperties') {
+      clonedState = { ...clonedState, ...obj.extraData };
+      history.push({ ...clonedState });
+    }
+
+    if (obj.type === 'removeProperties') {
+      for (const key of obj.keysToRemove) {
+        delete clonedState[key];
+      }
+
+      history.push({ ...clonedState });
+    }
```

Comment:
Instead of using multiple `if` statements, you can use a `switch` statement to handle different types of actions. This will make your code cleaner and easier to read. Also, remember to include a `default` case in your `switch` statement for error handling. [CODE STYLE]